### DISCUSSION TITLE
RTMP -> HLS -> S3

### DISCUSSION THREAD

**Discussion by ImNotSafeLLC** [bluenviron/mediamtx#2131] [Questions] [answered]
My goal is to have a user stream video via RTMP to mediamtx, which will then convert the stream to HLS and save it to an S3 bucket. The video should also be made available to be streamed live as well (from S3 - not mediamtx).
The HLS files will be removed from the mediamtx server after being stored in S3.

What I'm thinking:
Using _runOnReady_ and ffmpeg to convert RTMP to HLS.
Then running a /bin/ash script that uses inotifywait to detect closed HLS segment files - which then uses [s5cmd](https://github.com/peak/s5cmd) to push the files to S3.

Anyone else have experience with this?

**Comment by aler9** [comment]
Hello, this has been already answered here: https://github.com/bluenviron/mediamtx/discussions/1943#discussioncomment-6298544

> Regarding HLS packaging, keep in mind that nowadays serving a static folder through HTTP is not enough to generate a Low-latency HLS stream, since the latest specification introduced preloading paths, that are paths that keep clients on hold until a certain part or segment is available, and this feature can't achieved by serving a static folder. Therefore, my advice is to generate HLS by using MediaMTX instances in the cloud and use Amazon CloudFront in front of them to distribute the load, instead of using S3 buckets. MediaMTX provides a HTTP/HLS server that supports preloading paths and all LL-HLS features.

**Reply by ImNotSafeLLC** [reply]
My plan is to have MediaMTX be ephemeral to scale easier; it will auto scale up and down with demand via AWS ECS. Even having a single instance of MediaMTX on EC2 would risk data loss in event of a failure.
I looked into having CloudFront permanently cache/store the videos, but there's no guarantee.

I think _inotifywait_ + aws cli in a shell script could be a solution. Just wasn't sure if there were known best-practices out there.
`inotifywait --recursive --monitor --format '%w%f' --event close_write ./streams`

Side note: It looks like I may not be using MediaMTX for what it is intended. The regex path parser doesn't seem to work, which means we can only use statically named paths. I think this project looks very promising (lightweight, multi-platform, llhls) 🤩. I think I might take a look and see what's up with the regex path issue.

**Reply by aler9** [reply, marked as answer]
In my opinion you have to split the two tasks:

* regarding storage, you can use FFmpeg to dump streams to disk as described in the README (or keep an eye on the native feature here: #1399) and inotifywait

* regarding live HLS, you have to use a LL-HLS-capable server like MediaMTX, so use MediaMTX + CloudFront.

[a comment by an automated account is not shown]
